3. Вектор/Динамический массив

VitaliyIliyushenkov opened the pull request:
new file:   Vector_app/Vector_app/App.config
	new file:   Vector_app/Vector_app/Form1.Designer.cs
	new file:   Vector_app/Vector_app/Form1.cs
	new file:   Vector_app/Vector_app/Form1.resx
	new file:   Vector_app/Vector_app/MyVectEnum.cs
	new file:   Vector_app/Vector_app/MyVector.cs
	new file:   Vector_app/Vector_app/Program.cs
	new file:   Vector_app/Vector_app/Properties/AssemblyInfo.cs
	new file:   Vector_app/Vector_app/Properties/Resources.Designer.cs
	new file:   Vector_app/Vector_app/Properties/Resources.resx
	new file:   Vector_app/Vector_app/Properties/Settings.Designer.cs
	new file:   Vector_app/Vector_app/Properties/Settings.settings
	new file:   Vector_app/Vector_app/Vector_app.csproj

VitaliyIliyushenkov committed:
new file:   Vector_app/Vector_app.sln
	new file:   Vector_app/Vector_app/App.config
	new file:   Vector_app/Vector_app/Form1.Designer.cs
	new file:   Vector_app/Vector_app/Form1.cs
	new file:   Vector_app/Vector_app/Form1.resx
	new file:   Vector_app/Vector_app/MyVectEnum.cs
	new file:   Vector_app/Vector_app/MyVector.cs
	new file:   Vector_app/Vector_app/Program.cs
	new file:   Vector_app/Vector_app/Properties/AssemblyInfo.cs
	new file:   Vector_app/Vector_app/Properties/Resources.Designer.cs
	new file:   Vector_app/Vector_app/Properties/Resources.resx
	new file:   Vector_app/Vector_app/Properties/Settings.Designer.cs
	new file:   Vector_app/Vector_app/Properties/Settings.settings
	new file:   Vector_app/Vector_app/Vector_app.csproj

VasilevsWork commented on Vector_app/Vector_app/MyVectEnum.cs:
```diff
@@ -0,0 +1,44 @@
+            _data = data; this.size = size;
+        }
+        
+        public T Current { get { return _data[currentIndex]; } }
```
Не хватает `set`

VitaliyIliyushenkov replied:
public T Current { 
    get { return _data[currentIndex]; }
    set { _data[currentIndex] = value; }
}

VasilevsWork commented on Vector_app/Vector_app/MyVector.cs:
```diff
@@ -0,0 +1,88 @@
+           
+        }
+        
+        public T SetValue(int item)
```
Метод должен присваивать значение элементу с указанным индексом. Соответственно должно быть два параметра. А ничего возвращать не требуется.

VitaliyIliyushenkov replied:
Не правильно понял задание, исправлено.

VasilevsWork commented on Vector_app/Vector_app/MyVector.cs:
```diff
@@ -0,0 +1,88 @@
+
+namespace Vector_app
+{ 
+    public class MyVector<T> : IEnumerable<T>
```
Отсутствует переопределение индексатора (`[]`)

VitaliyIliyushenkov replied:
public T this[int index]
{
    get { return _data[index]; }
    set { _data[index] = value;}
}

VasilevsWork reviewed (commented):
Ряд мелких недочетов, однако их следует поправить. В остальном все хорошо.

VitaliyIliyushenkov committed:
modified:   Vector_app/Vector_app/Form1.Designer.cs
	modified:   Vector_app/Vector_app/Form1.cs
	modified:   Vector_app/Vector_app/MyVectEnum.cs
	modified:   Vector_app/Vector_app/MyVector.cs

VasilevsWork commented on Vector_app/Vector_app/MyVector.cs:
```diff
@@ -0,0 +1,94 @@
+            if (newSize <= 0) throw new Exception("Неверное число");
+
+            if (newSize < Size) Size = newSize;
+            Capacity = newSize;
```
`Resize` меняет текущий размер вектора. Делать `Capacity` таким же, как и `Size` в большинстве случаев неэффективно. Стоит менять делать `Capacity = newSize * 2;`.

VitaliyIliyushenkov replied:
Изменил на Capacity = newSize * 2;

VasilevsWork commented on Vector_app/Vector_app/MyVector.cs:
```diff
@@ -0,0 +1,94 @@
+
+        public T this[int index]
+        {
+            get { return _data[index]; }
+            set { _data[index] = value;}
```
Хорошо, что добавили индексатор. Однако следует использовать уже реализованные методы `At` и `SetValue`.

VitaliyIliyushenkov replied:
public T this[int index]
 {
     get { return At(index); }
     set { SetValue(index, value); }
 }

 public T At(int index) 
 {
     return _data[index]; 
 }

public void SetValue(int index, T item)
{
    if(index < 0 ||  index > Size) throw new Exception("Индекс вне массива");
    _data[index] = item;
}

VasilevsWork commented on Vector_app/Vector_app/MyVector.cs:
```diff
@@ -0,0 +1,94 @@
+            Capacity = newSize;
+        }
+
+        public void Reserve(int newCapacity)
```
Метод не меняет размер вектора, только вместимость. Вместимость при этом только увеличивается.

VitaliyIliyushenkov replied:
public void Reserve(int newCapacity)
{
    if (newCapacity <= 0 || newCapacity < Size)
        throw new Exception("Неверное число");
    else Capacity = newCapacity;
}

VasilevsWork reviewed (approved):
Общая реализация приемлемая, хоть и требуется еще ряд доработок. Задание принимается, однако очень желательно исправить замечания.

VitaliyIliyushenkov committed:
Последние доработки MyVector.cs
